Extend HDF5 output for yt

forrestglines opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

Adds some meta data and AMR level information to the HDF5 outputs so that the python analysis library `yt` can read it. Development of the companion `yt` frontend is at https://github.com/forrestglines/yt/tree/athenapk-frontend

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [ ] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md
- [ ] (@lanl.gov employees) Update copyright on changed files

Yurlungur commented:
@forrestglines will you submit the yt frontend as a PR to yt when it's ready?

Yurlungur commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -307,7 +307,85 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+                  H5S::FromHIDCheck(H5Screate_simple(1, xDims, NULL)), myDSet);  
+    }
+
+    { // TODO(forrestglines) - should this data be confined to yt-Info
+      // Store yt-necessary attributes
+      H5S const localDSpace = H5S::FromHIDCheck(H5Screate(H5S_SCALAR));
+      H5D const ytInfo =
```
I actually see no reason to tag this as "needed for yt." Who knows, other vis tools might find this output useful.

forrestglines replied:
After writing the prototype, I agree. It's a bit convoluted to wrap these into a separate dataset and should belong in the `Info` dataset with everything else. I think we should always write this data rather than create two classes of outputs - one that's readable by `yt` (or other codes) and one that's not

Yurlungur commented on src/outputs/parthenon_hdf5.cpp:
Let's make sure the yt reader is robust against the absence of this stuff

forrestglines commented:
> @forrestglines will you submit the yt frontend as a PR to yt when it's ready?

I'll ask the `yt-dev` mailing list about their standard for inclusion. I would be in favor of adding it.

Yurlungur commented:
My experience is they're very inclusive.

forrestglines commented:
Here's my current design for downstream hooks for labeling individual components of variables and including parameters as attributes in the HDF5 file.

To label individual components, I've added an optional `std::vector<std:string> component_labels_` to the `Metadata` class. This vector is used to name components along the first dimension of a variable - `"Density"`,`"MomentumDensity1"`, etc. for AthenaPK. If this vector is not specified, the variable is treated as one component with the same name as the variable label.

 I explored adding this vector to the `StateDescriptor` or to `CellVariable`, but it turned out to affect too many lines of code.
  
To add parameters to the HDF5 file, I added a `std::vector<std::string> outputing_params_` to `StateDescriptor`, which is used as a simple list of parameters already in `Params`.  I also added a optional `bool outputing` to `StateDescriptor::AddParam`, which defaults to `false`, to mark if a parameter should be included with the HDF5 file. Right now, only `float`, `double`, and `std:string` parameters are writable to HDF5.

See [this](https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/blob/forrestglines/hdf5-for-yt/src/hydro/hydro.cpp#L161) for how I add component labels in AthenaPK and [here](https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/blob/forrestglines/hdf5-for-yt/src/hydro/hydro.cpp#L134) for how I add `AdiabaticIndex` and `EquationOfState` as parameters for the HDF5 file.

forrestglines commented:
Do I need to add to the `xml` in the `xdmf` file? The `xdmf` file is not used within `yt`

Yurlungur commented:
> Here's my current design for downstream hooks for labeling individual components of variables and including parameters as attributes in the HDF5 file.
> 
> To label individual components, I've added an optional `std::vector<std:string> component_labels_` to the `Metadata` class. This vector is used to name components along the first dimension of a variable - `"Density"`,`"MomentumDensity1"`, etc. for AthenaPK. If this vector is not specified, the variable is treated as one component with the same name as the variable label.
> 
> I explored adding this vector to the `StateDescriptor` or to `CellVariable`, but it turned out to affect too many lines of code.
> 
> To add parameters to the HDF5 file, I added a `std::vector<std::string> outputing_params_` to `StateDescriptor`, which is used as a simple list of parameters already in `Params`. I also added a optional `bool outputing` to `StateDescriptor::AddParam`, which defaults to `false`, to mark if a parameter should be included with the HDF5 file. Right now, only `float`, `double`, and `std:string` parameters are writable to HDF5.
> 
> See [this](https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/blob/forrestglines/hdf5-for-yt/src/hydro/hydro.cpp#L161) for how I add component labels in AthenaPK and [here](https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/blob/forrestglines/hdf5-for-yt/src/hydro/hydro.cpp#L134) for how I add `AdiabaticIndex` and `EquationOfState` as parameters for the HDF5 file.

This seems like a reasonable solution. Can the output name for a variable be different than the code name? i.e., is there a way to translate variable names to the canonical yt names?

> Do I need to add to the xml in the xdmf file? The xdmf file is not used within yt

Hmm... I don't think so... I think this information can be handed to paraview separately... Although a paraview expert like @nmsriram is perhaps better equipped to answer.

forrestglines commented:
> This seems like a reasonable solution. Can the output name for a variable be different than the code name? i.e., is there a way to translate variable names to the canonical yt names?

Yes, these names are only defined for HDF5 outputs for `yt`. These labels are independent of any variable names already in use.

forrestglines commented:
@par-hermes format

ghost commented:
<img src="https://www.deepcode.ai/icons/green_check.svg" width= "50px" align= "left"/> Congratulations :tada:. DeepCode [analyzed](https://www.deepcode.ai/app/gh/lanl/parthenon/6151e40db99c16efac893ef9a71972df6f841008/lanl/parthenon/4c1c0cdca27ab805f81d54e6c361c65a22bb5dc4/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=0&) your code in 2.215 seconds and we found no issues. Enjoy a moment of no bugs :sunny:.

#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/lanl/parthenon/6151e40db99c16efac893ef9a71972df6f841008/lanl/parthenon/4c1c0cdca27ab805f81d54e6c361c65a22bb5dc4/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=0&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=lanl)

JoshuaSBrown commented:
Is this ready for review?

forrestglines commented:
@JoshuaSBrown It was ready - but then I decided to extend the testing so that the HDF5 comparison checks the metadata of the HDF5 files as well.

That extension should be ready today or tomorrow but we'll need to rebless the gold data as well.

forrestglines commented:
The extension to HDF5 metadata testing is finished. The python scripts now test that both files contain the same data in `Info`, allowing exceptions like differences in blocks per processing element and the order of mesh blocks in the file.

We do need to rebless the all the HDF5 gold data since metadata has been added to these files. However, this PR is ready for review.

JoshuaSBrown commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+auto pkg = std::make_shared<StateDescriptor>("Hydro");
+
+/* ... */
+
+std::vector<std::string> cons_labels(nhydro);
````
I would add:
```
const int nhydro = 5;
```

JoshuaSBrown commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+cons_labels[2]="MomentumDensity2";
+cons_labels[3]="MomentumDensity3";
+cons_labels[4]="TotalEnergyDensity";
+Metadata m({Metadata::Cell, Metadata::Independent, Metadata::FillGhost},
+           std::vector<int>({5}), cons_labels);
````
Remove magic number and use I'm guessing ```nhydro```?

JoshuaSBrown commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+           std::vector<int>({5}), cons_labels);
+pkg->AddField("cons", m);
+
+std::vector<std::string> density_labels(1);
+density_labels[0]="Density";
+m = Metadata({Metadata::Cell, Metadata::Derived}, std::vector<int>({1}), density_labels);
+pkg->AddField("dens", m);
+
+std::vector<std::string> velocity_labels(3);
+velocity_labels[0]="Velocity1";
+velocity_labels[1]="Velocity2";
+velocity_labels[2]="Velocity3";
+m = Metadata({Metadata::Cell, Metadata::Derived}, std::vector<int>({3}), velocity_labels);
+pkg->AddField("vel", m);
+
+std::vector<std::string> pressure_labels(1);
+pressure_labels[0]="Pressure";
+m = Metadata({Metadata::Cell, Metadata::Derived}, std::vector<int>({1}), pressure_labels);
+pkg->AddField("pres", m);
+```
````
Maybe I'm being a bit picky, but I would recommend for the purpose of being explicit removing the magic numbers and setting them equal to a value in each of these cases.

JoshuaSBrown commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+```float```, and ```std::string``` parameters can be included with the HDF5.
+
+Defining ```CodeLength```, ```CodeMass```, and ```CodeTime```, all in CGS,
+allows ```yt``` to intupret units for the data. The above example declares the
````
interpret

JoshuaSBrown commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+Defining ```CodeLength```, ```CodeMass```, and ```CodeTime```, all in CGS,
+allows ```yt``` to intupret units for the data. The above example declares the
+data as being in meters, kilograms, and seconds. Alternatively, this unit
+information can be also supplied to the ```yt``` frontend when loading the
````
can also be

Yurlungur commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+[yt](https://yt-project.org/) as certain variables are named during when adding
+fields via ```StateDescriptor::AddField```. Variable names are added as a
+```std::vector<std::string>``` in the variable metadata. These labels are only
+used for output to HDF5. 4D variables are named with a list of names for each
````
Are 5D variables supported? Are they simply rasterized?

forrestglines replied:
Currently not supported. The frontend is built around identifying either the conserved or primitive variables in  3D space, either being packed together in one 4D array, split up into individual 3D arrays, or some combination.

I'm not sure how we should interpret 5D data. Should the extra dimension be different species (`yt` as a whole doesn't really deal with multifluid) or should we interpolate them?

Yurlungur replied:
I was more thinking about, e.g., tensor quantities. i.e., the Reynolds stress for example. But multiple species (which I think we might implement as sparse variables) would be another example.

forrestglines replied:
Unfortunately, `yt` doesn't have any way to navigate tensors seemlessly. Each component would have to be named. From the way the output by Parthenon is implement and how I've implemented the `yt` frontend works currently, I think that would work without little to no modification to the `yt` frontend. 

We can keep that in mind if we want a separate `yt` test in `Parthenon` once the `yt` frontend is firmly solidified.

Yurlungur commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+
+std::vector<std::string> density_labels(1);
+density_labels[0]="Density";
+m = Metadata({Metadata::Cell, Metadata::Derived}, std::vector<int>({1}), density_labels);
````
Are these labels optional? Or are they now required?

forrestglines replied:
They're optional. If they're not specified, the HDF5 output will label the entire field by the field name, i.e. `cons` and `dens` here. I'll update the documentation.

Yurlungur commented on docs/outputs.md:
````diff
@@ -48,3 +48,105 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+pkg->AddField("pres", m);
+```
+
+For the ```yt``` frontend needs either the hydrodynamic conserved variables or
````
```suggestion
The `yt` frontend needs either the hydrodynamic conserved variables or
```

Yurlungur commented on src/interface/params.hpp:
```diff
@@ -41,15 +42,15 @@ class Params {
     PARTHENON_REQUIRE_THROWS(!(hasKey(key)), "Key " + key + " already exists");
     myParams_[key] = std::unique_ptr<Params::base_t>(new object_t<T>(value));
-    myTypes_[key] = std::string(typeid(value).name());
+    myTypes_.insert(make_pair(key, std::type_index(typeid(value))));
```
This seems like an improvement! But out of curiosity, why did it need to be changed now?

forrestglines replied:
Two reasons:

1. I changed the `myTypes_` map to hold `std::type_index` instead of `std::string` in order to be more sure of some type checking. This change likely wasn't needed for what I was trying to accomplish, but I imagine comparing `std::type_index` is safer than comparing strings.
2. The `[]` operator for maps (I think just for assignment) requires the type to have a default constructor. However, `std::type_index` doesn't have a default constructor since there isn't a default type. Hence using `insert` instead.

JoshuaSBrown replied:
Could you use emplace instead instead of insert.

JoshuaSBrown replied:
?

Yurlungur commented on src/interface/state_descriptor.hpp:
```diff
@@ -49,8 +49,11 @@ class StateDescriptor {
 
   template <typename T>
-  void AddParam(const std::string &key, T value) {
+  void AddParam(const std::string &key, T value, const bool outputing = false) {
```
The name `outputing` feels odd here. Why not just `output`?

forrestglines replied:
I used `outputing` to match the `outputing_params` vector that I've added. `OutputParams` is an entirely separate object, so I avoided using `output` as an verb here to avoid confusion. Would `include_with_output` be too verbose or should I just use `output`?

Yurlungur replied:
I like `include_with_output`

Yurlungur reviewed (approved):
Great work @forrestglines . Overall looks good. I have a few nitpicks below.

What's the status of the yt reader being in yt mainline?

forrestglines commented:
> What's the status of the yt reader being in yt mainline?

I have a fork for the `yt` frontend working, but it might change once we start doing analysis and include magnetic fields in our simulations. When we're sure that we have everything in the yt frontend working, I'll push that frontend to the yt mainline.

JoshuaSBrown commented on scripts/python/phdf.py:
```diff
@@ -110,24 +110,33 @@ def __init__(self, filename):
+            if '/VolumeLocations/x' in f:
+                self.x = f['/VolumeLocations/x'][:,:]
+            else:
+                self.x = np.zeros((self.NumBlocks,self.MeshBlockSize[0]))
+                for bId in range(self.NumBlocks):
+                    for cId in range(self.MeshBlockSize[0]):
+                        self.x[bId,cId] = 0.5*(tmpx[bId,cId]+tmpx[bId,cId+1])
```
I think you could put this logic into a function and pass in an index, and simply return the the numpy array. This would reduce the repetition below.

JoshuaSBrown commented on scripts/python/phdf_diff.py:
```diff
@@ -110,12 +110,43 @@ def compare(files, all=False, brief=True, quiet=False, one=False, tol=1.0e-12):
         return(3)
 
+    if check_meta_data:
+        #Check in the metadata in Info is identical, except use tol for comparing "Time"
+        f0_Info = { key:value for key,value in f0.Info.items() if key != "Time" and key != "BlocksPerPE" }
+        f1_Info = { key:value for key,value in f1.Info.items() if key != "Time" and key != "BlocksPerPE" }
+
+        errTime = np.abs(f0.Time- f1.Time)
+        if errTime > tol:
+            print("""
+            Time of outputs differ
+
+            Quitting...
+            """)
+            return(5)
+        if f0_Info.keys() != f1_Info.keys():
+            print("""
+            Attributes in '/Info' of outputs differ
+
+            Quitting...
+            """)
+            return(6)
+
+        info_diffs = list(np.any(f0_Info[key] != f1_Info[key]) for key in f0_Info.keys())
+        infos_differ = np.any( info_diffs )
+        if infos_differ:
+            print("\n'/Info' of outputs differ\n")
+            print("Differing attributes: ",list( (key) for key in f0_Info.keys() if np.any(f0_Info[key] != f1_Info[key]))  )
+            print("\nQuitting...\n")
+            return(7)
```
I would move this logic into a separate method, but that is more preference.

JoshuaSBrown commented on scripts/python/phdf_diff.py:
```diff
@@ -110,12 +110,43 @@ def compare(files, all=False, brief=True, quiet=False, one=False, tol=1.0e-12):
+
+        errTime = np.abs(f0.Time- f1.Time)
+        if errTime > tol:
+            print("""
+            Time of outputs differ
+
+            Quitting...
+            """)
```
Because it is floating point I would recommend actually printing the difference in the error message.

JoshuaSBrown commented on scripts/python/phdf_diff.py:
```diff
@@ -136,9 +167,50 @@ def compare(files, all=False, brief=True, quiet=False, one=False, tol=1.0e-12):
     for var in f0.Variables:
-        if var == 'Locations' or var == 'Info':
+
+        if var in ['Locations', 'VolumeLocations', 'Info', 'LogicalLocations', 'Levels']:
+            if check_meta_data and var != 'Info':
+                if var in ['Locations', 'VolumeLocations']:
+                    for key in f0.fid[var].keys():
+                        #Compare raw data of these variables
+                        val0 = f0.fid[var][key]
+                        val1 = f1.fid[var][key]
+
+                        #Sort val1 by otherBlockIdx
+                        val1 = val1[otherBlockIdx]
+
+                        # Compute norm error, check against tolerance
+                        errVal = np.abs(val0 - val1)
+                        errMag = np.linalg.norm(errVal)
+                        if(errMag > tol):
+                            no_meta_variables_diff = False
+                            if not quiet: print("")
+                            print(f'Metavariable {var}/{key} differs between {f0.file} and {f1.file}')
+                            if not quiet: print("")
+                        else:
+                            print('  %18s/%s: no diffs'%(var,key))
+                if var in ['LogicalLocations', 'Levels']:
+                    #Compare raw data of these variables
+                    val0 = f0.fid[var]
+                    val1 = f1.fid[var]
+
+                    #Sort val1 by otherBlockIdx
+                    val1 = val1[otherBlockIdx]
+
+                    #As integers, they should be identical
+                    if val0 != val0:
+                        no_meta_variables_diff = False
+                        if not quiet: print("")
+                        print(f'Metavariable {var} differs between {f0.file} and {f1.file}')
+                        if not quiet: print("")
+                    else:
+                        print('  %20s: no diffs'%var)
```
Again I would recommend packaging this block of logic into a separate method, ... preference. 
for floating point message I would print the difference.

JoshuaSBrown commented on scripts/python/phdf_diff.py:
```diff
@@ -204,6 +276,9 @@ def compare(files, all=False, brief=True, quiet=False, one=False, tol=1.0e-12):
             break
 
+    if not no_meta_variables_diff:
+        return(8)
```
Are the return values defined somewhere in the documentation or am I simply missing them.

forrestglines replied:
It looks like these return values were made sequential starting from 1, using the standard return value of 0 to the command line to indicate no errors. I don't think the nonzero return values are treated specially anywhere. Would enumerating the return values at the top of the file be sufficient?

JoshuaSBrown replied:
Yes, I think so, simply having a summary in the docstring of what each exit code means would be nice so you don't have to go digging through the logic.

JoshuaSBrown commented on src/interface/metadata.hpp:
```diff
@@ -346,13 +353,17 @@ class Metadata {
   void Associate(const std::string &name) { associated_ = name; }
   const std::string &getAssociated() const { return associated_; }
 
+  const std::vector<std::string> getComponentLabels() const { return component_labels_; }
```
This method can be made noexcept.

JoshuaSBrown commented on src/interface/params.hpp:
```diff
@@ -41,15 +42,15 @@ class Params {
   void Update(const std::string &key, T value) {
     PARTHENON_REQUIRE_THROWS((hasKey(key)), "Key " + key + "missing.");
-    PARTHENON_REQUIRE_THROWS(!(myTypes_.at(key).compare(std::string(typeid(T).name()))),
+    PARTHENON_REQUIRE_THROWS(myTypes_.at(key) == std::type_index(typeid(T)),
```
This is clearer, nice!

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -372,14 +461,99 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+      // Write Levels and Logical Locations with the level for each Meshblock
+      // loclist contains levels and logical locations for all meshblocks on
+      // all ranks
+      std::vector<LogicalLocation> loclist = pm->GetLocList();
```
Can  you make this a const reference, doesn't look like you are altering it's contents.

JoshuaSBrown commented on src/interface/state_descriptor.hpp:
```diff
@@ -67,10 +70,17 @@ class StateDescriptor {
   // infers type correctly.
   template <typename T>
   const T &Param(const std::string &key, T value) {
-    params_.Get(key, value);
+    return params_.Get(key, value);
+  }
+
+  const std::type_index &ParamType(const std::string &key) {
+    return params_.GetType(key);
   }
```
These methods look like they should be made const.

JoshuaSBrown commented on src/interface/state_descriptor.hpp:
```diff
@@ -67,10 +70,17 @@ class StateDescriptor {
 
   Params &AllParams() { return params_; }
+
+  std::vector<std::string> &IncludeWithOutputParams() { return includeWithOutputParams_; }
+
```
Can be made const noexcept.

JoshuaSBrown replied:
It looks like you are also only adding to includeWithOutputParams_ from within the Statedescriptor class so you could probably return by const reference as well.

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -274,45 +279,129 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+                  H5S::FromHIDCheck(H5Screate_simple(1, xDims, NULL)), infoDSet);
+
+      // RootGridDomain - float[9] array with xyz mins, maxs, rats (dx(i)/dx(i-1))
+      Real rootGridDomain[9] = {
+          pm->mesh_size.x1min, pm->mesh_size.x1max, pm->mesh_size.x1rat,
+          pm->mesh_size.x2min, pm->mesh_size.x2max, pm->mesh_size.x2rat,
+          pm->mesh_size.x3min, pm->mesh_size.x3max, pm->mesh_size.x3rat};
+      const hsize_t rootGridDomain_size[1] = {9};
+      writeH5AF64("RootGridDomain", rootGridDomain,
+                  H5S::FromHIDCheck(H5Screate_simple(1, rootGridDomain_size, NULL)),
+                  infoDSet);
+
+      // RootGridSize - int[3] number of cells on the root grid
+      int rootGridSize[3] = {pm->mesh_size.nx1, pm->mesh_size.nx2, pm->mesh_size.nx3};
+      const hsize_t rootGridSize_size[1] = {3};
```
I would reverse the definition of rootGridDomain_size and rootGridSize_size so they appear before rootGridDomain and rootGridSize respectively, that way you can use them when defining both structures and get rid of some magic numbers.

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -274,45 +279,129 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+                  infoDSet);
+
+      // BoundaryConditions
+      std::vector<std::string> boundaryConditions(BOUNDARY_NFACES);
+      for (int i = 0; i < BOUNDARY_NFACES; i++) {
+        boundaryConditions[i] = GetBoundaryString(pm->mesh_bcs[i]);
+      }
```
In general I believe it will be more performant to call reserve with push_back because you get to avoid a default construction call. Not that it will really matter here.

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -274,45 +279,129 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+
+        if (component_labels.size() > 0) {
+          numVariables[i] = component_labels.size();
+          for (auto &label : component_labels) {
```
make const auto

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -274,45 +279,129 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+            double param = state->Param<double>(param_name);
+            writeH5AF64(ss.str().c_str(), &param, localDSpace, infoDSet);
+          } else if (type == typeid(std::string)) {
+            std::string param = state->Param<std::string>(param_name);
```
can param be made a const reference?

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -274,45 +279,129 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+          ss << state->label() << "/" << param_name;
+
+          if (type == typeid(int)) {
+            int param = state->Param<int>(param_name);
+            writeH5AI32(ss.str().c_str(), &param, localDSpace, infoDSet);
+          } else if (type == typeid(double)) {
+            double param = state->Param<double>(param_name);
```
make param const?

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -274,45 +279,129 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+        std::shared_ptr<StateDescriptor> state = package.second;
+        // Write IncludeWithOutputParams
+        std::vector<std::string> outputing_params = state->IncludeWithOutputParams();
+        for (auto &param_name : outputing_params) {
```
make param_name const reference

JoshuaSBrown commented on src/outputs/parthenon_hdf5.hpp:
```diff
@@ -256,6 +256,27 @@ static void writeH5ASTRING(const char *name, const std::string &pData,
+static void writeH5ASTRINGS(const char *name, const std::vector<std::string> &pData,
+                            const hid_t &dSpace, const hid_t &dSet) {
+  int max_name_length = 0;
+  for (auto s : pData) {
```
make s const reference.

JoshuaSBrown commented on src/outputs/parthenon_hdf5.hpp:
```diff
@@ -256,6 +256,27 @@ static void writeH5ASTRING(const char *name, const std::string &pData,
+    max_name_length = (max_name_length > s.length()) ? max_name_length : s.length();
+  }
+
+  std::vector<const char *> c_strs;
```
call reserve after declaring c_strs;

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -372,14 +461,99 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+      std::vector<LogicalLocation> loclist = pm->GetLocList();
+
+      std::vector<std::int64_t> levels(pm->nbtotal);
+      std::vector<std::int64_t> logicalLocations(pm->nbtotal * 3);
```
call reserve instead and use with push_back.

JoshuaSBrown reviewed (changes requested):
Thanks for doing this @forrestglines this is a nice piece of work, the only thing that I would really like to see is some tests that account for the extra output.

forrestglines commented:
@par-hermes format

forrestglines commented:
> Thanks for doing this @forrestglines this is a nice piece of work, the only thing that I would really like to see is some tests that account for the extra output.

The new metadata checking in the python scripts checks for the new metadata added, assuming that the gold data has those extra attributes.

The one new piece that isn't exercised in the testing are the naming of variable components via `AddField` and marking parameters for output via `AddParam`. I could add that to the advection example and that functionality would be covered.

forrestglines commented:
@par-hermes format

JoshuaSBrown commented on scripts/python/phdf.py:
```diff
@@ -109,35 +109,23 @@ def __init__(self, filename):
-                    for cId in range(self.MeshBlockSize[2]):
-                        self.z[bId,cId] = 0.5*(tmpz[bId,cId]+tmpz[bId,cId+1])
-            self.zf = tmpz
+            def load_coord(coord_i):
+                coord_name = ["x","y","z"][coord_i]
+
+                tmp = f[f'/Locations/{coord_name}'][:,:]
+                if f'/VolumeLocations/{coord_name}' in f:
+                    coord = f[f'/VolumeLocations/{coord_name}'][:,:]
+                else:
+                    coord = np.zeros((self.NumBlocks,self.MeshBlockSize[coord_i]))
+                    for bId in range(self.NumBlocks):
+                        for cId in range(self.MeshBlockSize[coord_i]):
+                            coord[bId,cId] = 0.5*(tmp[bId,cId]+tmp[bId,cId+1])
+                coordf=tmp
+                return tmp,coord,coordf
+
+            tmpx, self.x, self.xf = load_coord(0)
+            tmpy, self.y, self.yf = load_coord(1)
+            tmpz, self.z, self.zf = load_coord(2)
```
Beautiful :)

JoshuaSBrown commented on scripts/python/phdf_diff.py:
```diff
@@ -64,7 +64,90 @@ def addPath():
+    errTime = np.abs(f0.Time- f1.Time)
+    if errTime > tol:
+        print(f"""
+        Time of outputs differ by {f0.Time - f1.Time}
+
+        Quitting...
```
👍🏻

forrestglines commented:
@JoshuaSBrown I've covered all the changes you requested. I also added the new metadata functionality to the advection example so that it gets tested during regression testing along with the rest of the HDF5 output.

forrestglines commented:
This PR is ready to merge except for final approval and reblessing of the gold data for HDF5 outputs. I don't know how to go about reblessing.

jlippuner commented on docs/outputs.md:
````diff
@@ -59,3 +59,108 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+## Preparing outputs for ```yt```
+
+Parthenon HDF5 outputs can be read with the python visualization library
+[yt](https://yt-project.org/) as certain variables are named during when adding
````
```suggestion
[yt](https://yt-project.org/) as certain variables are named when adding
```

jlippuner commented on docs/outputs.md:
````diff
@@ -59,3 +59,108 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+pkg->AddParam<int>("IntParam", 0,true);
+pkg->AddParam<std::string>("EquationOfState", "Adiabatic",true);
+```
+all function to add parameters to the HDF5 output. Currently, only ```int```,
````
I don't understand "all function to add parameters to the HDF5 output." here

forrestglines replied:
I'll change to "add the parameters `CodeLength`, `CodeMass`, `CodeTime`, `AdiabaticIndex`, `IntParam`, and `EquationOfState` to the HDF5 output" since it's more descriptive.

jlippuner commented on docs/outputs.md:
````diff
@@ -59,3 +59,108 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+
+Defining ```CodeLength```, ```CodeMass```, and ```CodeTime```, all in CGS,
+allows ```yt``` to interpret units for the data. The above example declares the
+data as being in meters, kilograms, and seconds. Alternatively, this unit
````
I am confused. The above example doesn't specify units explicitly, so wouldn't that mean they're in centimeters, grams, seconds by default? Where are meters and kilograms coming from?

forrestglines replied:
You specify `CodeLength`, `CodeMass`, and `Code Time` in terms of centimeters in a code length, grams in a code mass, and seconds in a code time. So in the example you have 100 cm in a code length (1 code length is 1 m ), 1000 g in a code mass (1 code mass is 1 kg), and 1 s in a code time (1 code time is 1 s). I'll make that more clear

jlippuner commented on scripts/python/phdf.py:
```diff
@@ -109,26 +109,23 @@ def __init__(self, filename):
+            def load_coord(coord_i):
+                coord_name = ["x","y","z"][coord_i]
+
+                tmp = f[f'/Locations/{coord_name}'][:,:]
+                if f'/VolumeLocations/{coord_name}' in f:
+                    coord = f[f'/VolumeLocations/{coord_name}'][:,:]
```
Hmm, the fact that `f` is used as the file and as the prefix for formatting strings is a bit confusing here... Can we maybe do this?
```suggestion
                tmp = f['/Locations/' + coord_name][:,:]
                vol_loc = '/VolumeLocations/' + coord_name
                if vol_loc in f:
                    coord = f[vol_loc][:,:]
```

forrestglines replied:
That's fine here. Usually I avoid concatenations with strings and prefer the python string formatting, but here it is confusing with the file `f`.

jlippuner commented on scripts/python/phdf_diff.py:
```diff
@@ -64,8 +66,121 @@ def addPath():
+
+    otherBlockIdx = list(f0.findBlockIdxInOther(f1,i) for i in range(f0.NumBlocks))
+
+    for var in set(f0.Variables+f1.Variables):
+        if var in ['Locations', 'VolumeLocations', 'LogicalLocations', 'Levels']:
+            if var in ['Locations', 'VolumeLocations']:
```
This seems unnecessarily complicated to me. Can't we just make the first block `for var in ['Locations', 'VolumeLocations']:` and the second block `for var in ['LogicalLocations', 'Levels']:` and then have a check that `var` exists in the files? That seems more readable to me.

forrestglines replied:
Yes. This was the result of this code block moving around.

jlippuner commented on src/interface/metadata.hpp:
```diff
@@ -181,6 +181,13 @@ class Metadata {
   }
 
+  /// returns a metadata with bits and shape set
+  explicit Metadata(const std::vector<MetadataFlag> &bits, const std::vector<int> &shape,
+                    const std::vector<std::string> component_labels)
+      : shape_(shape), sparse_id_(-1), component_labels_(component_labels) {
```
Are there any consistency rules between the `shape` and the number of `component_labels` that we should/could enforce?

forrestglines replied:
Not any simple rules right now. Currently, the `yt` interface would accept any number of `component_labels` as long as it could reshape the raw array into right `nlabels*nz*ny*nx` shape. Since we're not sure at the moment how downstream codes will arrange or label their components, I'm not sure there's a simple rule that we can impose at the moment.

jlippuner commented on src/interface/params.hpp:
```diff
@@ -83,12 +84,18 @@ class Params {
+  const std::type_index &GetType(const std::string &key) const {
+    auto const it = myParams_.find(key);
+    PARTHENON_REQUIRE_THROWS(it != myParams_.end(), "Key " + key + " doesn't exist");
+    return myTypes_.at(key);
```
```suggestion
    return it->second;
```

jlippuner commented on src/interface/state_descriptor.hpp:
```diff
@@ -49,8 +49,11 @@ class StateDescriptor {
 
   template <typename T>
-  void AddParam(const std::string &key, T value) {
+  void AddParam(const std::string &key, T value, const bool include_with_output = false) {
```
Is there a reason to not just include all parameters with the output and not give the user an option?

forrestglines replied:
That would be an interesting idea. We don't have that many parameters, so it wouldn't be that much of a file size change to write them all to disk. It might clutter the output a bit but it would simplify the code.

The one obstacle is that currently we don't have methods to write all types of parameters as an HDF5 attribute. The simple integer, floating point, and string types are easy to write, but more complicated ones like the equation of state class that we have in AthenaPK.

I think it would make sense to write all simple parameters in a separate `Params` dataset as long as we're ok with all that data in the HDF5 file. We could limit it to simple types for now and in the future let parameter types implement a `WriteAsHDF5Attribute` if they also need to be added to the HDF5 file. We don't have any use cases for that though.

For this PR I can make all `int`, `double`, and `std::string`  parameters write as an attribute and in your PR we can add other simple types.

jlippuner replied:
I realized this later. Yes, I think just writing all params with simple type to the outputfile and ignoring complex types is the right way to go.

jlippuner commented on src/interface/state_descriptor.hpp:
```diff
@@ -249,6 +261,8 @@ class StateDescriptor {
   Dictionary<Metadata> swarmMetadataMap_;
   Dictionary<Dictionary<Metadata>> swarmValueMetadataMap_;
+
+  std::vector<std::string> includeWithOutputParams_;
```
Maybe use a `std::set` to avoid duplicates? Or get rid of this altogether and just write out all params, as suggested above.

forrestglines replied:
Removing this vector with the new commit.

jlippuner commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -274,45 +279,132 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+          std::stringstream ss;
+          ss << state->label() << "/" << param_name;
+
+          if (type == typeid(int)) {
+            const int &param = state->Param<int>(param_name);
+            writeH5AI32(ss.str().c_str(), &param, localDSpace, infoDSet);
+          } else if (type == typeid(double)) {
+            const double &param = state->Param<double>(param_name);
+            writeH5AF64(ss.str().c_str(), &param, localDSpace, infoDSet);
+          } else if (type == typeid(std::string)) {
+            const std::string &param = state->Param<std::string>(param_name);
+            writeH5ASTRING(ss.str().c_str(), param, localDSpace, infoDSet);
+          }
```
We should probably have `uint32_t`, `uint64_t`, `int64_t`, and `float` here as well. This will become a bit easier with my PR #486 that makes writing attributes a template function on the type.

forrestglines replied:
Yes. I only added the types that already had a convenient macro to write the attribute. In your PR with the template function, we can add the other obvious types.

jlippuner replied:
I'm fine with this.

jlippuner commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -372,14 +464,102 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+
+      // Only write levels on rank 0
+      local_count[0] = (Globals::my_rank == 0) ? pm->nbtotal : 0;
+      local_count[1] = 0;
+      local_count[2] = 0;
+      local_count[3] = 0;
+      local_count[4] = 0;
```
This is not necessary (same for `global_count`) since only the first entry will be ready when we create a rank 1 space

jlippuner commented on src/outputs/parthenon_hdf5.hpp:
```diff
@@ -256,6 +256,28 @@ static void writeH5ASTRING(const char *name, const std::string &pData,
+                            const hid_t &dSpace, const hid_t &dSet) {
+  int max_name_length = 0;
+  for (const auto &s : pData) {
+    max_name_length = (max_name_length > s.length()) ? max_name_length : s.length();
```
Should use `std::max` here, but it appears that `max_name_length` doesn't get used. I wouldn't worry about this function, though, because I have a similar, more general, function to write vector attributes in my PR #486, so I'll remove this in that PR.

jlippuner reviewed (approved):
Nice work! This is a good addition to the code to allow easy plotting of Parthenon output with yt.

I added some questions and nitpicks, but nothing blocking.

JoshuaSBrown commented on docs/outputs.md:
````diff
@@ -59,3 +59,108 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+m = Metadata({Metadata::Cell, Metadata::Derived}, std::vector<int>({nvelocity}), velocity_labels);
+pkg->AddField("vel", m);
+
+const in npressure = 1;
````
```suggestion
const int npressure = 1;
```

JoshuaSBrown commented on docs/outputs.md:
````diff
@@ -59,3 +59,108 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+ds = yt.load(filename,units_override=units_override,gamma=5./3.)
+```
+
+Currently the ```yt``` frontend for Parthenon is hosted on the
````
```suggestion
Currently, the ```yt``` frontend for Parthenon is hosted on the
```

JoshuaSBrown commented on scripts/python/phdf_diff.py:
```diff
@@ -26,15 +26,16 @@
-         -brief: Only report if files are different
-                 Overrides --all
-         -quiet: Only report if files are different and
-                 don't print any extraneous info.
-      --tol=eps: set tolerance to eps.  Default 1.0e-12
+                  -all: report all diffs at all positions
+                  -one: Quit after first different variable
+                -brief: Only report if files are different
+                        Overrides --all
+                -quiet: Only report if files are different and
+                        don't print any extraneous info.
+             --tol=eps: set tolerance to eps.  Default 1.0e-12
+      -ignore_metadata: Ignore differences in metadata
```
I don't expect any changes, but do you know why some of these are using double dashes and others are not is this a convention? I always thought the double dashes were for the long version and single dashes for short versions of the same argument, it looks like here the double dash prefix is being used only in cases where an assignment is involved. Could probably create an issue (low priority), this should probably be uniform between our python scripts.

forrestglines replied:
I also don't see any consistency here. I think `./tst/regression/run_test.py` is the only other python script to do argument parsing.

pgrete replied:
I also think consistency would be nice and also not necessary for this PR -> "good first issue"

JoshuaSBrown commented on src/interface/state_descriptor.hpp:
```diff
@@ -49,8 +49,11 @@ class StateDescriptor {
 
   template <typename T>
-  void AddParam(const std::string &key, T value) {
+  void AddParam(const std::string &key, T value, const bool include_with_output = false) {
```
If you do end up using the boolean, I would recommend changing it to an enum class which would be more descriptive:

```
enum class IncludeWithOutput {
    True,
    False
};

AddParam<std::string>("Parthenon is Awesome", IncludeWithOutput::True);
```

jlippuner replied:
That's a good suggestion. Although it would be a non-issue if we just output all parameters of simple type as discussed above.

forrestglines replied:
I'm removing the boolean, otherwise it's a good suggestion.

jlippuner commented on docs/outputs.md:
````diff
@@ -59,3 +59,114 @@ which will produce a ```png``` image per dump suitable for encoding into a movie
+units used by Parthenon in terms of centimeters, grams, and seconds by writing
+the parameters.  In the above example, these parameters dictate ```yt``` to
+interpret code lengths in the data in units of 100 centimeters (or 1 meter per
+code unit), code masses in units of 1000 grams ( or 1 kilogram per code units)
+and code times in units of seconds ( or 1 second per code time).
````
Thanks, I get it now :) Two small fixes:
```suggestion
code unit), code masses in units of 1000 grams (or 1 kilogram per code units)
and code times in units of seconds (or 1 second per code time).
```

pgrete commented on scripts/python/phdf.py:
```diff
@@ -109,26 +109,24 @@ def __init__(self, filename):
+                if vol_loc in f:
+                    coord = f[vol_loc][:,:]
+                else:
+                    coord = np.zeros((self.NumBlocks,self.MeshBlockSize[coord_i]))
+                    for bId in range(self.NumBlocks):
+                        for cId in range(self.MeshBlockSize[coord_i]):
+                            coord[bId,cId] = 0.5*(tmp[bId,cId]+tmp[bId,cId+1])
```
Wouldn't sth like
```suggestion
                    coord = 0.5*(tmp[:,:-1] + tmp[:,1:])
```
also work here as both for loops cover the entire range?

pgrete replied:
Also, may be worth adding a warning here (printed) that this assume uniform Cartesian grids so that we don't miss it when adding other coordinate systems.

pgrete commented on scripts/python/phdf_diff.py:
```diff
@@ -64,8 +66,120 @@ def addPath():
+    ERROR_TIME_DIFF=10
+    ERROR_INFO_ATTRS_DIFF=12
+    ERROR_INFO_VALUES_DIFF=13
+    ERROR_META_VARS_DIFF=14
```
```suggestion
    ERROR_INFO_ATTRS_DIFF=11
    ERROR_INFO_VALUES_DIFF=12
    ERROR_META_VARS_DIFF=13
```

Probably a leftover from an earlier iteration, but the indices between doc and code don't match.

forrestglines replied:
Good catch, I just fixed this

pgrete commented on scripts/python/phdf_diff.py:
```diff
@@ -64,8 +66,120 @@ def addPath():
+
+    errTime = np.abs(f0.Time- f1.Time)
+    if errTime > tol:
+        print(f"""
```
Nice, f-strings, I approve ;)

pgrete commented on scripts/python/phdf_diff.py:
```diff
@@ -64,8 +66,120 @@ def addPath():
+                val1 = f1.fid[var][key]
+
+                #Sort val1 by otherBlockIdx
+                val1 = val1[otherBlockIdx]
```
Leaving a note for myself here to follow up on what's happening here.
Haven't seen that kind of indexing with a list yet.

forrestglines replied:
`otherBlockIdx` gives the indexing of mesh blocks on `f1` in the same order as the meshblocks in `f0`, since meshblocks are free to rearrange in order due to changing MPI ranks.

pgrete commented on scripts/python/phdf_diff.py:
```diff
@@ -64,8 +66,120 @@ def addPath():
+            val1 = val1[otherBlockIdx]
+
+            #As integers, they should be identical
+            if val0 != val0:
```
```suggestion
            if val0 != val1:
```
Also, does this like work without an `.any()` ?

forrestglines replied:
It should be `np.any( val0 != val1)` for full correctness. It might be checking equivalence by reference here

pgrete commented on scripts/python/phdf_diff.py:
```diff
@@ -64,8 +66,120 @@ def addPath():
+    ERROR_NO_OPEN_F0 = 1
+    ERROR_NO_OPEN_F1 = 2
+    ERROR_CELLS_DIFFER = 3
+    ERROR_DATA_DIFFER = 4
```
:+1: I like this.

forrestglines commented:
@par-hermes format

pgrete reviewed (approved):
Nice work! I like the flexible design through the state descriptor.
My major comments mostly concern the Python piece which may be worth double checking.

pgrete commented:
Additional note on updating the "gold standard":

1. In an environment where current `develop` passes create the new gold standard and put into a single file named `parthenon_regression_gold_v#.tgz`  with `#` being the "next" number (here 4)
2. Go to https://github.com/lanl/parthenon/releases and create a new release with the file as binary attachment and branch name `regression-gold-v#`, cf., e.g., https://github.com/lanl/parthenon/releases/tag/regression-gold-v3
3. Edit the root `CMakeLists.txt` (in this PR) and update both number and SHA512 hash:
```
set(REGRESSION_GOLD_STANDARD_VER 3 CACHE STRING "Version of gold standard to download and use")
set(REGRESSION_GOLD_STANDARD_HASH
  "SHA512=2445000a031aafc9a85a684aa7e82fb5ccda05530c5652ab432a8fb254642c7b18252ff55d8306b1120aba898713712b28b804eb30272b06278698046a6461cc"
  CACHE STRING "Hash of default gold standard file to download")
```
4. IIRC that's it: Fingers crossed ;)

forrestglines commented:
@pgrete Thanks! I'll rebless the tests and merge once the CI passes in this PR.

forrestglines committed:
Added metadata to HDF5 output for yt

forrestglines commented:
>     2\. Go to https://github.com/lanl/parthenon/releases and create a new release with the file as binary attachment and branch name `regression-gold-v#`, cf., e.g., https://github.com/lanl/parthenon/releases/tag/regression-gold-v3

So do I use the tag `regression-gold-v4` and target branch `develop` or the branch in this PR?

forrestglines committed:
Updated gold standard for HDF5 metadata

forrestglines committed:
Added  <algorithm> for style check
